Point_set_processing_3/include/CGAL/estimate_lfs.h

```diff
@@ -0,0 +1,687 @@
+  FT lip_lby = s.y() - lipschitz * (lip_lbx - s.x());
+
+  assert(lip_lbx + eps >= s.x() - eps);
+  assert(lip_lbx - eps <= t.x() + eps);
+
+  FT lip_ubx = ((t.y() - s.y()) / lipschitz + t.x() + s.x()) * 0.5;
+  FT lip_uby = s.y() + lipschitz * (lip_ubx - s.x());
+
+  assert(lip_ubx + eps >= s.x() - eps);
+  assert(lip_ubx - eps <= t.x() + eps);
+
+  // stop
+  if (lip_lby >= epsilon_band || lip_uby <= epsilon_band)
+  {
+    return true;
+  }
+
+  const bool s_sign = (s.y() > epsilon_band);
+  const bool t_sign = (t.y() > epsilon_band);
+  const FT s_slope = (s_sign) ? -lipschitz : lipschitz;
+  const FT t_slope = (t_sign) ? lipschitz : -lipschitz;
+
+  FT ms_x = s.x() + (epsilon_band - s.y()) / s_slope;
+  const Point pms = ps + ms_x / len * v;
+  FT ms_y = classical_point_dist_func(pms, neighbor_query);
+  const Point_2 ms(ms_x, ms_y);
+  assert(ms_x + eps >= s.x() - eps);
+
+  FT mt_x = t.x() + (epsilon_band - t.y()) / t_slope;
+  const Point pmt = ps + mt_x / len * v;
+  FT mt_y = classical_point_dist_func(pmt, neighbor_query);
+  const Point_2 mt(mt_x, mt_y);
+  assert(mt_x - eps <= t.x() + eps);
+  assert(mt_x + eps >= ms_x - eps);
+
+  const FT mm_x = (ms_x + mt_x) / 2;
+  const Point pmm = ps + mm_x / len * v;
+  FT mm_y = classical_point_dist_func(pmm, neighbor_query);
+  Point_2 mm(mm_x, mm_y);
+
+  if (std::abs(mm_y - epsilon_band) <= eps && (mt_x - ms_x) <= epsilon_band / lipschitz)
+  {
+    if (s_sign == t_sign)
+    {
+            intersections.push_back(pmm);
+            intersections.push_back(pmm);
+    }
+    else
+    {
+      intersections.push_back(pmm);
+    }
+        
+      return true;
+  }
+
+  bool fl = recursive_dichotomic_search_base(e, ms, mm, neighbor_query, epsilon_band, 
+            intersections, lipschitz, eps);
+  bool fr = recursive_dichotomic_search_base(e, mm, mt, neighbor_query, epsilon_band, 
+            intersections, lipschitz, eps);
+
+  return (fl && fr);
+}
+
+template<class NeighborQuery>
+bool recursive_dichotomic_search(const typename NeighborQuery::Kernel::Segment_3 &e, 
+                                const typename NeighborQuery::Kernel::Point_2 &s, 
+                                const typename NeighborQuery::Kernel::Point_2 &t,
+                                const NeighborQuery &neighbor_query,
+                                const typename NeighborQuery::FT epsilon_band, 
+                                std::vector<typename NeighborQuery::Kernel::Point_3>& intersections,
+                                typename NeighborQuery::FT lipschitz, 
+                                typename NeighborQuery::FT eps=1e-5)
+{
+  // basic geometric types
+  typedef typename NeighborQuery::Kernel Kernel;
+  typedef typename Kernel::Point_3  Point;
+
+  CGAL_precondition(intersections.empty());
+  bool flag = recursive_dichotomic_search_base(e, s, t, neighbor_query, epsilon_band, 
+                    intersections, lipschitz, eps);
+
+  if (flag == false) // should not false
+  {
+    ;
```

Comment:
????

Reply:
this is used for debugging. i will remove it.